### PR TITLE
MiniCanvas needs early code-review, this isn't a real pull request

### PR DESCRIPTION
**Pull request by cobbal**
The idea:

mini-canvas is an http server that sumulates the endpoints of the canvas API. It's similar to CoreUITest's mock functions, but different in a few ways

## CoreUITest.mock(...)
 - works on exact URL matches, no parsing of parameters
 - used per test case
 - can't tell the difference between a GET or a DELETE (hasn't really been an issue so far)

## MiniCanvas
 - can parse both path and query parameters. e.g. /api/v1/courses/:courseID?include[]=that_thing
 - endpoints are fixed, but the state of the server can be easily changed and reset
   + This will make mocked tests more reliable, as outstanding requests from a previous test will no longer cause a missing mock failure to occur on the next test 
 - no failing tests for missing mocks, just returns a 404
 - used (hopefully) globally, with test cases making small tweaks if needed
 - actually is a real HTTP server

For now, all the important code is in MiniCanvasServer.swift which is only ~250 lines so far. I will definitely break the file up before trying to merge it.

To test the server I've been running "FigureOutWhereToPutThis" in nightly tests.

### CODE REVIEW

**Comment by inst-danger** [comment]
## Latest Installable Builds

(will redirect to this page if no builds exist yet)
student: [Link](https://rebrand.ly/a4qyvt2) | [QR](https://qr.rebrandly.com/v1/qrcode?shortUrl=rebrand%2Ely%2Fa4qyvt2)
teacher: [Link](https://rebrand.ly/358g2iq) | [QR](https://qr.rebrandly.com/v1/qrcode?shortUrl=rebrand%2Ely%2F358g2iq)
parent: [Link](https://rebrand.ly/lng6qie) | [QR](https://qr.rebrandly.com/v1/qrcode?shortUrl=rebrand%2Ely%2Flng6qie)

File: Core/Core.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -4192,17 +4192,15 @@
...
-				"$(SRCROOT)/../Carthage/Build/iOS/PSPDFKit.framework",
-				"$(SRCROOT)/../Carthage/Build/iOS/PSPDFKitUI.framework",
 			);
 			name = "Carthage Copy Frameworks";
 			outputFileListPaths = (
+				"$(SRCROOT)/../scripts/carthage/output.xcfilelist",
 			);
 			outputPaths = (
-				"$(BUILT_PRODUCTS_DIR)/$(FRAMEWORKS_FOLDER_PATH)/PSPDFKit.framework",
-				"$(BUILT_PRODUCTS_DIR)/$(FRAMEWORKS_FOLDER_PATH)/PSPDFKitUI.framework",
```
Comment (vanwagonet) [inline Core/Core.xcodeproj/project.pbxproj]: wonder if this will help with our duplicate errors.

Comment (cobbal) [reply Core/Core.xcodeproj/project.pbxproj]: I don't think so. I managed to track the duplicate thing down to a difference in HEADER_SEARCH_PATH in the binary pcm files, but I couldn't seem to make this change cause that.

**Comment by inst-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  Please add a release note. If no release note is wanted, use `none`. Example: `release note: Fixed a bug that prevented users from enjoying the app.`
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  Please add which apps this change affects. Example: `affects: Teacher, Student` or `affects: none`
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  Please add a reference to a jira ticket. For example: `refs: MBL-10023`
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  Build failed, skipping coverage check
  </td>
    </tr>
  </tbody>
</table>



<details><summary>:x: XCTest failed: CoreTests/APIOAuthRequestableTests/testGetMobileVerifyRequest</summary>


    Assertion Failure: APIOAuthTests.swift:24: XCTAssertEqual failed: ("http://localhost:8000/api/v1/mobile_verify.json") is not equal to ("https://canvas.instructure.com/api/v1/mobile_verify.json")
</details>
<details><summary>:x: XCTest failed: ParentUnitTests/AddStudentControllerTests/testAddPairingCode</summary>


    Crash: Parent (15980): ParentTestCase.router.getter
</details>

<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

**Pull request closed without merging**